system/security/LdapSecurity/ldapsecurity.cpp

```diff
@@ -654,32 +656,68 @@ bool CLdapSecManager::authenticate(ISecUser* user)
         return false;
     }
 
-    if(user->getAuthenticateStatus() == AS_AUTHENTICATED)
-        return true;
+    user->setAuthenticateStatus(AS_UNKNOWN);
 
-    if(m_permissionsCache->isCacheEnabled() && !m_usercache_off && m_permissionsCache->lookup(*user))
+    bool isCaching = m_permissionsCache->isCacheEnabled() && !m_usercache_off;//caching enabled?
+    bool isUserCached = false;
+    Owned<ISecUser> cachedUser = new CLdapSecUser(user->getName(), "");
+    if(isCaching)
     {
-        user->setAuthenticateStatus(AS_AUTHENTICATED);
-        return true;
+        user->copyTo(*(cachedUser.get()));//copy user to cachedUser
+        isUserCached = m_permissionsCache->lookup(*cachedUser);//populate cachedUser with cached values
     }
 
-    if ((user->credentials().getSessionToken() != 0) || !isEmptyString(user->credentials().getSignature()))//Already authenticated it token or signature exist
+    //Verify provided signature if present
+    IDigitalSignatureManager * pDSM = createDigitalSignatureManagerInstanceFromEnv();
+    if (pDSM && pDSM->isDigiVerifierConfigured() && !isEmptyString(user->credentials().getSignature()))
+    {
+        StringBuffer b64Signature(user->credentials().getSignature());
+        if (!pDSM->digiVerify(user->getName(), b64Signature))//digital signature valid?
+        {
+            user->setAuthenticateStatus(AS_INVALID_CREDENTIALS);
+            WARNLOG("Invalid digital signature for user %s", user->getName());
+            return false;
+        }
+        else
+        {
+            user->setAuthenticateStatus(AS_AUTHENTICATED);
+            if(isCaching)
+                m_permissionsCache->add(*user);
+            return true;
+        }
+    }
```

Comment:
Could the whole '//Verify provided signature if present' section be moved in the front of Line 661 since the section does not need the isCaching, isUserCached, and cachedUser?

Reply:
I need "isCaching" on line 684, which requires me to lookup the user in the cache.  However, I see that i need to check "isUserCached" test on line 684 as follows
if(isCaching && !isUserCached)